cmd/remote-configuration/agent/agent.go

```diff
@@ -0,0 +1,283 @@
+	if err != nil {
+		return err
+	}
+
+	agent.logger.Debugf("OpAMP Client started.")
+
+	return nil
+}
+
+func (agent *Agent) updateAgentIdentity(instanceId ulid.ULID) {
+	agent.logger.Debugf("Agent identify is being changed from id=%v to id=%v",
+		agent.instanceId.String(),
+		instanceId.String())
+	agent.instanceId = instanceId
+
+	// TODO: reinit or update meter (possibly using a single function to update all own connection settings
+}
+
+func (agent *Agent) getNameAndNamespace(key string) (string, string, error) {
+	s := strings.Split(key, "/")
+	// We expect map keys to be of the form name/namespace
+	if len(s) != 2 {
+		return "", "", errors.New("invalid key")
+	}
+	return s[0], s[1], nil
+}
+
+func (agent *Agent) makeKeyFromNameNamespace(name string, namespace string) string {
+	return fmt.Sprintf("%s/%s", name, namespace)
+}
+
+func (agent *Agent) getEffectiveConfig(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+	instances, err := agent.applier.ListInstances()
+	if err != nil {
+		agent.logger.Errorf("couldn't list instances", err)
+		return nil, err
+	}
+	instanceMap := map[string]*protobufs.AgentConfigFile{}
+	for _, instance := range instances {
+		marshaled, err := yaml.Marshal(instance)
+		if err != nil {
+			agent.logger.Errorf("couldn't marshal collector configuration", err)
+			return nil, err
+		}
+		mapKey := agent.makeKeyFromNameNamespace(instance.GetName(), instance.GetNamespace())
+		instanceMap[mapKey] = &protobufs.AgentConfigFile{
+			Body:        marshaled,
+			ContentType: "yaml",
+		}
+	}
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: instanceMap,
+		},
+	}, nil
+}
+
+func (agent *Agent) initMeter(settings *protobufs.TelemetryConnectionSettings) {
+	reporter, err := metrics.NewMetricReporter(agent.logger, settings, agent.config.GetAgentType(), agent.config.GetAgentVersion(), agent.instanceId)
+	if err != nil {
+		agent.logger.Errorf("Cannot collect metrics: %v", err)
+		return
+	}
+
+	if agent.metricReporter != nil {
+		agent.metricReporter.Shutdown()
+	}
+	agent.metricReporter = reporter
+}
+
+// Take the remote config, layer it over existing, done
+// INVARIANT: The caller must verify that config isn't nil _and_ the configuration has changed between calls.
+func (agent *Agent) applyRemoteConfig(config *protobufs.AgentRemoteConfig) (*protobufs.RemoteConfigStatus, error) {
+	var multiErr error
+	for key, file := range config.Config.GetConfigMap() {
+		if len(key) == 0 || len(file.Body) == 0 {
+			continue
+		}
+		name, namespace, err := agent.getNameAndNamespace(key)
+		if err != nil {
+			multiErr = multierr.Append(multiErr, err)
+			continue
+		}
+		err = agent.applier.Apply(name, namespace, file)
+		if err != nil {
+			multiErr = multierr.Append(multiErr, err)
+			continue
+		}
+		agent.appliedKeys[key] = true
+	}
+	for collectorKey := range agent.appliedKeys {
+		name, namespace, err := agent.getNameAndNamespace(collectorKey)
```

Comment:
I don't think getting the name and namespace is a lot of work (and it's only added to `appliedKeys` if this function works anyways), but couldn't we save them in `appliedKeys` so we don't have to keep doing this?